### PR TITLE
support collecting network records per pass

### PR DESCRIPTION
**Pull request by brendankenny**
- always records network information (fixing #604)
- allows recording the network activity for multiple passes, paving the way for audits like #578
- warns you if your current config has conflicting names for its trace data
- renames `traceName` in config files to the more accurate `passName`

This does use the same naming method as the earlier trace bucket PR. We still need to find a different solution, but at least tracing and network will be unified again, making them easier to change together.

### CODE REVIEW

File: lighthouse-core/audits/critical-request-chains.js [hunk 1]
```diff
@@ -40,7 +40,8 @@ class CriticalRequestChains extends Audit {
    * @return {!AuditResult} The score from the audit, ranging from 0-100.
    */
   static audit(artifacts) {
-    return artifacts.requestCriticalRequestChains(artifacts.networkRecords).then(chains => {
+    const networkRecord = artifacts.networkRecords[Audit.DEFAULT_TRACE];
```
Comment (paulirish) [inline lighthouse-core/audits/critical-request-chains.js]: probably need to rename it to `DEFAULT_PASS` as this looks odd..

**Comment by brendankenny** [comment]
updated. Fixes, updated tests, and new tests added

**Comment by brendankenny** [comment]
also renamed `traceName` in the config file to `passName` to be more accurate. This would be a breaking change, but I don't believe anyone uses a named pass yet

File: lighthouse-core/audits/critical-request-chains.js [hunk 2]
```diff
@@ -40,7 +40,8 @@ class CriticalRequestChains extends Audit {
    * @return {!AuditResult} The score from the audit, ranging from 0-100.
    */
   static audit(artifacts) {
-    return artifacts.requestCriticalRequestChains(artifacts.networkRecords).then(chains => {
+    const networkRecord = artifacts.networkRecords[Audit.DEFAULT_PASS];
```
Comment (paulirish) [inline lighthouse-core/audits/critical-request-chains.js]: should be plural

File: lighthouse-core/gather/gather-runner.js [hunk 3]
```diff
@@ -227,10 +224,13 @@ class GatherRunner {
...
-            .then(loadData => {
-              // Merge pass trace and network data into tracingData.
-              config.trace && Object.assign(tracingData.traces, loadData.traces);
-              config.network && (tracingData.networkRecords = loadData.networkRecords);
+            .then(passData => {
+              // If requested by config, merge trace and network data for this
+              // pass into tracingData.
+              const passName = config.passName || Audit.DEFAULT_PASS;
+              config.trace && (tracingData.traces[passName] = passData.trace);
+              config.network && (tracingData.networkRecords[passName] = passData.networkRecords);
```
Comment (paulirish) [inline lighthouse-core/gather/gather-runner.js]: in a readthough it's not clear that network is recorded but not saved. Should we rename the config property to `config.saveNetwork` or similar to clarify?

File: lighthouse-core/gather/gather-runner.js [hunk 4]
```diff
@@ -135,7 +136,8 @@ class GatherRunner {
     const driver = options.driver;
     const config = options.config;
     const gatherers = config.gatherers;
-    const loadData = {traces: {}};
+    const loadData = {};
```
Comment (paulirish) [inline lighthouse-core/gather/gather-runner.js]: rename to passData

File: lighthouse-core/lib/log.js [hunk 5]
```diff
@@ -39,13 +39,23 @@ function _log(title, logargs) {
...
       this.emit(title, args);
     }
   }
+
+  /**
+   * Fires off all warnings. Listen with
+   * `require('lib/log').events.addListener('warning', callback)`
+   */
+  issueWarning(args) {
+    this.emit('warning', args);
```
Comment (paulirish) [inline lighthouse-core/lib/log.js]: how about `...args`? I think the other evt handlers distribute the args

File: lighthouse-core/test/gather/gather-runner-test.js [hunk 6]
```diff
@@ -173,27 +184,7 @@ describe('GatherRunner', function() {
 
     return GatherRunner.afterPass({driver, config}).then(vals => {
```
Comment (paulirish) [inline lighthouse-core/test/gather/gather-runner-test.js]: vals -> passData

File: lighthouse-core/test/gather/gather-runner-test.js [hunk 7]
```diff
@@ -349,15 +365,16 @@ describe('GatherRunner', function() {
...
+      passName: 'firstPass',
       loadPage: true,
       gatherers: [new TestGatherer()]
     }];
     const options = {driver: fakeDriver, url: 'https://example.com', flags: {}, config: {}};
 
     return GatherRunner.run(passes, options)
         .then(artifacts => {
-          const p = artifacts.requestCriticalRequestChains(artifacts.networkRecords);
+          const networkRecords = artifacts.networkRecords.firstPass;
```
Comment (paulirish) [inline lighthouse-core/test/gather/gather-runner-test.js]: const networkRecords = artifacts.networkRecords['firstPass'];

File: lighthouse-core/lib/log.js (hunk 5 above)
Comment (paulirish) [inline lighthouse-core/lib/log.js]: nvm

**Comment by paulirish** [comment]
lgtm % nits

File: lighthouse-core/gather/gather-runner.js [hunk 8]
```diff
@@ -144,32 +146,24 @@ class GatherRunner {
...
-            traceEvents: traceContents
-          };
-        }
-
-        const traceName = config.traceName || Audit.DEFAULT_TRACE;
-        loadData.traces[traceName] = traceContents;
-        loadData.traceEvents = traceContents.traceEvents;
+        // traceContents was an array of trace events; after, traceContents is
+        // an object with a traceEvents property. Normalize to object form.
+        loadData.trace = Array.isArray(traceContents) ? {
```
Comment (brendankenny) [inline lighthouse-core/gather/gather-runner.js]: it just barely doesn't fit

File: lighthouse-core/gather/gather-runner.js (hunk 3 above)
Comment (brendankenny) [inline lighthouse-core/gather/gather-runner.js]: I feel like the implementation details aren't important, just that if a pass has a true `config.network

Comment (brendankenny) [inline lighthouse-core/gather/gather-runner.js]: I feel like the implementation details aren't important, just that if a pass has a true `config.network`, one is provided. I can put a note on top, though

Comment (paulirish) [inline lighthouse-core/gather/gather-runner.js]: note on top wfm

**Comment by brendankenny** [comment]
PTAL

**Pull request merged**